software/apps/systolic/matmul/main.c

```diff
@@ -0,0 +1,186 @@
+  mempool_barrier_init(core_id);
+
+  // Initialization
+  mempool_init(core_id, num_cores);
+
+  // Allocate systolic grid mapping
+  if (core_id == 0) {
+    grid_mapping = (uint32_t *)simple_malloc(num_cores * 4);
+    if (num_cores != SYSTOLIC_SIZE * SYSTOLIC_SIZE) {
+      printf("SYSTOLIC_SIZE does not match core count!\n");
+      return -1;
+    }
+  }
+
+  // Assign grid position (row wise)
+  // uint32_t col_idx = core_id % SYSTOLIC_SIZE;
+  // uint32_t row_idx = core_id / SYSTOLIC_SIZE;
+
+  // Assign grid position (col wise)
+  uint32_t col_idx = core_id / SYSTOLIC_SIZE;
+  uint32_t row_idx = core_id % SYSTOLIC_SIZE;
+
+  // 16 CORES only
+  // Assign grid position (tile wise)
+  // uint32_t col_idx;
+  // uint32_t row_idx;
+  // if (core_id < 4) {
+  //   col_idx = core_id % 2;
+  //   row_idx = core_id / 2;
+  // } else if (core_id < 8) {
+  //   col_idx = core_id % 2 + 2;
+  //   row_idx = core_id / 6;
+  // } else if (core_id < 12) {
+  //   col_idx = core_id % 2;
+  //   row_idx = core_id / 10 + 2;
+  // } else {
+  //   col_idx = core_id % 2 + 2;
+  //   row_idx = core_id / 14 + 2;
+  // }
+
+  // Wait for all cores
+  mempool_barrier(num_cores);
+
+  // Set systolic grid mapping
+  grid_mapping[row_idx * SYSTOLIC_SIZE + col_idx] = tile_id;
+
+  // Wait for all cores
+  mempool_barrier(num_cores);
+
+  // Setup
+  if (core_id == 0) {
+    printf("> Initialize\n");
+
+    // Print out grid mapping
+    // print_matrix((int32_t *)grid_mapping, 16, 16);
+
+    // Initialize systolic array
+    systolic_init(grid_mapping);
+
+    // Create systolic matrices
+    generate_gradient_matrix(&matrix_A, DIM_M, DIM_N);
+    systolic_matrix_create(&syst_matrix_A, matrix_A, DIM_M, DIM_N);
+    simple_free(matrix_A);
+    generate_gradient_matrix(&matrix_B, DIM_N, DIM_P);
+    systolic_matrix_create(&syst_matrix_B, matrix_B, DIM_N, DIM_P);
+    simple_free(matrix_B);
+    systolic_matrix_allocate(&syst_matrix_C, DIM_M, DIM_P);
+
+    // Print out systolic matrices A & B
+    // printf("> Print Systolic Matrices A & B\n");
+    // systolic_matrix_print(syst_matrix_A);
+    // systolic_matrix_print(syst_matrix_B);
+
+    // Set repetition count per submatrix of C (A->num_cols == B->num_rows)
+    rep_count = syst_matrix_A->num_cols / 2;
+  }
+
+  // Wait for all cores
+  mempool_barrier(num_cores);
+
+  if (core_id == 0) {
+    // Start benchmark
+    printf("> Start\n");
+    mempool_start_benchmark();
+  }
+
+  // Wait for all cores
+  mempool_barrier(num_cores);
+
+  if ((row_idx == 0) && (col_idx == 0)) {
+    systolic_rcp_pe(rep_count, syst_matrix_A, syst_matrix_B, syst_matrix_C);
+  }
+
+  if ((row_idx == 0) && (col_idx != 0)) {
+    systolic_cp_pe(col_idx, rep_count, syst_matrix_B, syst_matrix_C);
+  }
+
+  if ((row_idx != 0) && (col_idx == 0)) {
+    systolic_rp_pe(row_idx, rep_count, syst_matrix_A, syst_matrix_C);
+  }
+
+  if ((row_idx != 0) && (col_idx != 0)) {
+    systolic_np_pe(row_idx, col_idx, rep_count, syst_matrix_C);
+  }
```

Comment:
Would be nice if we had a comment about the different roles that the PEs take, depending on where they are in the CGRA.